### PR TITLE
change to epytope 3.3.1 version and new multi container for peptide prediction

### PR DESCRIPTION
**Pull request by jonasscheid**
Changes:

- New multi container has been created since mulled container did not work
-> https://github.com/users/jonasscheid/packages/container/package/epitopeprediction-2

- Other modules have been updated to epytope 3.3.0

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/epitopeprediction/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/epitopeprediction _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jonasscheid** [comment]
```
File "/opt/conda/lib/python3.7/site-packages/epytope/IO/MartsAdapter.py", line 158, in __perform_request
[166](https://github.com/nf-core/epitopeprediction/actions/runs/4873272243/jobs/8692613955?pr=200#step:4:167)
      backoff_factor=3
[167](https://github.com/nf-core/epitopeprediction/actions/runs/4873272243/jobs/8692613955?pr=200#step:4:168)
  TypeError: __init__() got an unexpected keyword argument 'allowed_methods'
```

Hmm the marts adapter seems to have problems. Any ideas @christopher-mohr

**Comment by christopher-mohr** [comment]
> ```
> File "/opt/conda/lib/python3.7/site-packages/epytope/IO/MartsAdapter.py", line 158, in __perform_request
> [166](https://github.com/nf-core/epitopeprediction/actions/runs/4873272243/jobs/8692613955?pr=200#step:4:167)
>       backoff_factor=3
> [167](https://github.com/nf-core/epitopeprediction/actions/runs/4873272243/jobs/8692613955?pr=200#step:4:168)
>   TypeError: __init__() got an unexpected keyword argument 'allowed_methods'
> ```
> 
> Hmm the marts adapter seems to have problems. Any ideas @christopher-mohr

I guess the `Requests` and/or `urllib3` package versions are too old.

**Comment by jonasscheid** [comment]
I still don't know how to fix this.. It seems to me that nextflow does not switch into the epytope environment within the docker container, even though that is happening when executing the container locally..

**Comment by christopher-mohr** [comment]
> I still don't know how to fix this.. It seems to me that nextflow does not switch into the epytope environment within the docker container, even though that is happening when executing the container locally..

I will have look again today.

**Comment by jonasscheid** [comment]
Good point 🤔  we actually could try a [python 3.7 container](https://github.com/docker-library/python/tree/master/3.7), pip install epytope and apt/yum-install gawk, tcsh and perl

**Comment by christopher-mohr** [comment]
> Good point 🤔 we actually could try a [python 3.7 container](https://github.com/docker-library/python/tree/master/3.7), pip install epytope and apt/yum-install gawk, tcsh and perl

Yes. that would have been my suggestion.

**Comment by christopher-mohr** [comment]
> After testing some containers I think I found one that works. Local tests for peptides and proteins runs through, however not test_full (with variants)
> 
> > Command error:
> > WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
> > Using TensorFlow backend.
> > INFO:**main**:Running Epitope Prediction And Annotation version: 1.1
> > INFO:**main**:Starting predictions at 2023-07-26 13:55:55
> > 2023-07-26 13:55:55,534 - **main** - INFO - Running Epitope Prediction And Annotation version: 1.1
> > 2023-07-26 13:55:55,534 - **main** - INFO - Starting predictions at 2023-07-26 13:55:55
> > ERROR:root:Attribute coding not found for dataset hsapiens_gene_ensembl on http://feb2014.archive.ensembl.org/biomart/martservice
> 
> Do we need to update biomart version in epytope again? I installed epytope 3.3.1

If it still works with the current pipeline version then I would suspect a change in the DB handling in combination with the database version that is used in the pipeline currently. I can have a look. Is the container also already online?

**Comment by jonasscheid** [comment]
Pushed new container.
`docker pull ghcr.io/jonasscheid/epitopeprediction-2:0.3.0`

**Comment by christopher-mohr** [comment]
This seems to be an issue that even exists on their page. When I try to query for the coding sequence using this archive version on their page I get the following:

```
ERROR: caught BioMart::Exception::Configuration: Attribute 'coding' not found in dataset default.hsapiens_gene_ensembl

If you repeatedly get directed to this error page, there may be a problem with your current session parameters. To clear your session and start with a clean slate, please click the New button below.

Stacktrace:
Exception::Class::Base::throw /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Registry.pm:490
BioMart::Registry::getAttribute /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Query.pm:1243
BioMart::Query::addAttribute /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Web.pm:808
BioMart::Web::prepare_martquery /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Web.pm:2129
BioMart::Web::handle_request /ensemblweb/archive/www_75/ensembl-webcode/biomart-perl/cgi-bin/martview:100
(eval) /ensemblweb/archive/www_75/ensembl-webcode/biomart-perl/cgi-bin/martview:99
ModPerl::ROOT::ModPerl::Registry::ensemblweb_archive_www_75_ensembl_2dwebcode_biomart_2dperl_cgi_2dbin_martview::handler /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:204
(eval) /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:204
ModPerl::RegistryCooker::run /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:170
ModPerl::RegistryCooker::default_handler /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/Registry.pm:31
ModPerl::Registry::handler -e:0
(eval) -e:0
```
I will ask them what's wrong.

File: modules/local/external_tools_import.nf [hunk 1]
```diff
@@ -52,7 +52,7 @@ process EXTERNAL_TOOLS_IMPORT {
     # Substitution 3: NMHOME should be the folder in which the tcsh script itself resides
     #
     sed -i.bak \
-        -e 's_bin/tcsh.*\$_usr/bin/env tcsh_' \
+        -e 's_bin/tcsh.*\$_opt/conda/bin/tcsh_' \
```
Comment (christopher-mohr) [inline modules/local/external_tools_import.nf:55]: Is this changed because of the container? Shouldn't this be container-unrelated?

**Comment by christopher-mohr** [comment]
> This seems to be an issue that even exists on their page. When I try to query for the coding sequence using this archive version on their page I get the following:
> 
> ```
> ERROR: caught BioMart::Exception::Configuration: Attribute 'coding' not found in dataset default.hsapiens_gene_ensembl
> 
> If you repeatedly get directed to this error page, there may be a problem with your current session parameters. To clear your session and start with a clean slate, please click the New button below.
> 
> Stacktrace:
> Exception::Class::Base::throw /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Registry.pm:490
> BioMart::Registry::getAttribute /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Query.pm:1243
> BioMart::Query::addAttribute /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Web.pm:808
> BioMart::Web::prepare_martquery /ensemblweb/archive/www_75/biomart-perl/lib/BioMart/Web.pm:2129
> BioMart::Web::handle_request /ensemblweb/archive/www_75/ensembl-webcode/biomart-perl/cgi-bin/martview:100
> (eval) /ensemblweb/archive/www_75/ensembl-webcode/biomart-perl/cgi-bin/martview:99
> ModPerl::ROOT::ModPerl::Registry::ensemblweb_archive_www_75_ensembl_2dwebcode_biomart_2dperl_cgi_2dbin_martview::handler /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:204
> (eval) /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:204
> ModPerl::RegistryCooker::run /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/RegistryCooker.pm:170
> ModPerl::RegistryCooker::default_handler /localsw/lib/perl5/site_perl/5.14.2/x86_64-linux-thread-multi/ModPerl/Registry.pm:31
> ModPerl::Registry::handler -e:0
> (eval) -e:0
> ```
> 
> I will ask them what's wrong.

They forwarded my request to the technical support team. Still waiting for feedback.

File: modules/local/external_tools_import.nf (hunk 1 above)
Comment (jonasscheid) [reply modules/local/external_tools_import.nf:55]: Good catch, that needs to be adjusted to the paths in the new container. That one was from the Conda-based container.

**Comment by jonasscheid** [comment]
I think I found the reason for the error

This is the selection list of variant types in the "older" archives.
Screenshot 2023-08-16 at 11 14 04

[since the 2018 archives ](https://www.ensembl.org/Help/ArchiveRedirect?src=http%3A%2F%2Fapr2018.archive.ensembl.org%2F) however, they use `Variant Consequence` as key for the selection list. And in these newer versions this error is not appearing anymore.

Screenshot 2023-08-16 at 11 24 59

So I guess they did not change this key for the older versions.

Off note: `coding` attribute can also not be found anymore on gene types (as far as I understood)

**Comment by christopher-mohr** [comment]
> I think I found the reason for the error
> 
> This is the selection list of variant types in the "older" archives. Screenshot 2023-08-16 at 11 14 04
> 
> [since the 2018 archives ](https://www.ensembl.org/Help/ArchiveRedirect?src=http%3A%2F%2Fapr2018.archive.ensembl.org%2F) however, they use `Variant Consequence` as key for the selection list. And in these newer versions this error is not appearing anymore.
> 
> Screenshot 2023-08-16 at 11 24 59
> 
> So I guess they did not change this key for the older versions.
> 
> Off note: `coding` attribute can also not be found anymore on gene types (as far as I understood)

That should not be the cause for the error we also get on their website. 

They also basically provide kind of a mapping of attribute names and respective database fields when using the API. So "Variant type" and "Variant Consequence" can still point to the same field. "coding" or "coding sequence" should be available in the "Sequence" category.

Btw I asked them for an update but was told that they are still looking into that.

**Comment by jonasscheid** [comment]
I see! Thanks for the explanation 🙏🏼 
Fingers crossed that they'll fix it soon

**Comment by christopher-mohr** [comment]
> I see! Thanks for the explanation 🙏🏼 Fingers crossed that they'll fix it soon

With the new `BiomartAdapter` in `epytope` we also make use of this mapping table that can be retrieved via API. So the database fields (that should be stable) can be mapped to the "names" that are needed for the queries. Therefore, we are flexible with respect to the version and should be more robust for future updates.

**Comment by christopher-mohr** [comment]
Changed the Ensembl reference version handling and fixed it for the most recent and most of the archive versions in https://github.com/nf-core/epitopeprediction/pull/213. I would suggest to close this PR in favor of https://github.com/nf-core/epitopeprediction/pull/213.

**Comment by jonasscheid** [comment]
Thanks for the update!

Yes, makes sense to me to close this PR 👍🏼  Finally then we have epytope 3.3.1 in there 🙌🏼

**Pull request closed without merging**